src/data/projectIdSingleton.ts

```diff
@@ -0,0 +1,114 @@
+    const { url, request } = _buildWhoamiRequest(environment, apiKey);
+
+    let response: Response;
+    try {
+      response = await fetch(url, request);
+    } catch (e: any) {
+      // Expected fetch exceptions listed here https://developer.mozilla.org/en-US/docs/Web/API/fetch#exceptions
+      // Most are header-related and should never occur since we do not let the user set headers. A TypeError
+      // will occur if the connection fails due to invalid environment configuration provided by the user. This is
+      // different from server errors handled below because the client is unable to make contact with a Pinecone
+      // server at all without a valid environment value.
+      if (e instanceof TypeError) {
+        throw new PineconeConfigurationError(
+          `A network error occured while attempting to connect to ${url}. Are you sure you passed the correct environment? Please check your configuration values and try again. Visit https://status.pinecone.io for overall service health information.`
+        );
+      } else {
+        throw new PineconeUnknownRequestFailure(url, e);
+      }
+    }
+
+    if (response.status >= 400) {
+      throw mapHttpStatusError({
+        status: response.status,
+        url,
+        message: await response.text(),
+      });
+    }
+
+    let json;
+    try {
+      json = await response.json();
+    } catch (e) {
+      throw new PineconeUnexpectedResponseError(
+        url,
+        response.status,
+        await response.text(),
+        'The HTTP call succeeded but the response could not be parsed as JSON.'
+      );
+    }
+
+    if (!json.project_name) {
+      throw new PineconeUnexpectedResponseError(
+        url,
+        response.status,
+        await response.text(),
+        'The HTTP call succeeded but response did not contain expected project_name.'
+      );
+    }
+
+    return json.project_name;
+  };
+
+  /** @hidden */
+  const _buildWhoamiRequest = (
+    environment: string,
+    apiKey: string
+  ): { url: string; request: RequestInit } => {
+    const url = `https://controller.${environment}.pinecone.io/actions/whoami`;
+    const request = {
+      method: 'GET',
+      headers: {
+        'Content-Type': 'application/json',
+        'Api-Key': apiKey,
+        'User-Agent': buildUserAgent(false),
+      },
+    };
+    return { url, request };
+  };
+
+  const key = (config) => `${config.apiKey}:${config.environment}`;
+
+  return {
+    getProjectId: async function (config: PineconeConfigurationOptions) {
+      const cacheKey = key(config);
+      if (cacheKey in projectIds) {
+        return projectIds[cacheKey];
+      } else {
+        const projectId = await _fetchProjectId(config);
+        projectIds[cacheKey] = projectId;
+        return projectId;
+      }
+    },
+
+    _reset: () => {
+      for (const key of Object.keys(projectIds)) {
+        delete projectIds[key];
+      }
+    },
+
+    _set: (config, projectId) => {
```

Comment:
The singleton is not exported from the module so everything in this file is "private" in that sense. The underscored methods are only for testing to cleanup state that would otherwise be a potential source of test pollution in the future (since the singleton is a singleton and doesn't get destroyed between test runs). The `getProjectId` (with no underscore) is used in actual production code. That's the distinction I was trying to create here. But none of these are really user-facing so I don't think it matters much.

Reply:
Got it, makes sense. Thanks!